### PR TITLE
[OPENSTACK-2966] add purge interface in driver

### PR DESCRIPTION
**Pull request by Niklaus-xie**

### CODE REVIEW

File: f5lbaasdriver/v2/bigip/driver_v2.py [hunk 1]
```diff
@@ -644,6 +644,54 @@ def refresh(self, context, body):
...
+                    device_loaded = self.driver.device_scheduler.load_device(
+                        context, device_id_passed
+                    )
+
+                    if device_loaded and device_loaded["admin_state_up"]:
+                        LOG.debug("using device %s to purge", device_id_passed)
+                        LOG.debug('real device in db: %s', device.get("id"))
+                        # to double check if this will delete unexpected items.
+                        service["device"] = device_loaded
+                    else:
```
Comment (zhaoqin-github) [inline f5lbaasdriver/v2/bigip/driver_v2.py:676]: I am not quite sure what http code will be returned to client, if this error is raised here, and I also not quite sure what state will this lb become to.

Ideally, if it is not allowed purge config upon disabled device, The api should return a reasonable http code, and lb provisioning state does not become to pending update, nor error.

Comment (zhaoqin-github) [reply f5lbaasdriver/v2/bigip/driver_v2.py:676]: However, I do not see inventory api does any validation for that.

Comment (Niklaus-xie) [reply f5lbaasdriver/v2/bigip/driver_v2.py:676]: code will be 409  . and a returned msg in the command line telling the caller that the device is not usable and operation unsuccessful. the status will be back to ACTIVE from Pending

File: f5lbaasdriver/v2/bigip/driver_v2.py [hunk 2]
```diff
@@ -644,6 +644,72 @@ def refresh(self, context, body):
...
+            LOG.error("Pls check its status. Purge unsuccessful here.")
+            # update its status to active here although it's not purged
+            self.driver.plugin.db.update_status(
+                context, models.LoadBalancer,
+                loadbalancer.id, plugin_constants.ACTIVE
+            )
+            raise
+
+        except Exception as e:
+            LOG.error("Exception: loadbalancer purge: %s" % e.message)
```
Comment (zhang-shengping) [inline f5lbaasdriver/v2/bigip/driver_v2.py:706]: do we need to use self._handle_entity_error?

File: f5lbaasdriver/v2/bigip/driver_v2.py [hunk 3]
```diff
@@ -644,6 +644,72 @@ def refresh(self, context, body):
...
+
+            if lbext.get("device", None):
+                device_id_passed = lbext["device"]
+                if device.get("id") != device_id_passed:
+                    LOG.warning('device passed from args. Trying to use it!')
+                    device_loaded = self.driver.device_scheduler.load_device(
+                        context, device_id_passed
+                    )
+
+                    if device_loaded and device_loaded["admin_state_up"]:
```
Comment (zhang-shengping) [inline f5lbaasdriver/v2/bigip/driver_v2.py:671]: do we need to consider the admin state down device?

**Review by zhang-shengping — APPROVED** [review]

File: f5lbaasdriver/v2/bigip/driver_v2.py (hunk 2 above)
Comment (Niklaus-xie) [reply f5lbaasdriver/v2/bigip/driver_v2.py:706]: the line below here(update_status) does what is similar to _handle_entity_error

File: f5lbaasdriver/v2/bigip/driver_v2.py (hunk 3 above)
Comment (zhaoqin-github) [reply f5lbaasdriver/v2/bigip/driver_v2.py:671]: Currently, admin_state_up == False can prevent cloud user to manipulate configuration against the device. However, admin operations, such as rebuild or purge, might need to manipulate the device when admin set this property False. We may need to consider that a little more.

Comment (Niklaus-xie) [reply f5lbaasdriver/v2/bigip/driver_v2.py:671]: https://github.com/F5Networks/f5-openstack-lbaasv2-driver/pull/1231 is an alternative way which does not check whether admin_state_up or down for device.

**Comment by Niklaus-xie** [comment]
using the way in https://github.com/F5Networks/f5-openstack-lbaasv2-driver/pull/1231 , closing this for now .

**Pull request closed without merging**